### PR TITLE
TSQL parser support for `(LHS) UNION RHS` queries

### LINKED ISSUE #1127: [TODO] Implement this style of UNION
See [more context](https://github.com/databrickslabs/remorph/blame/661b68471a043d68b673fbca50b31684fa447310/core/src/main/scala/com/databricks/labs/remorph/parsers/tsql/TSqlRelationBuilder.scala#L68-L73):
https://github.com/databrickslabs/remorph/blob/661b68471a043d68b673fbca50b31684fa447310/core/src/main/scala/com/databricks/labs/remorph/parsers/tsql/TSqlRelationBuilder.scala#L68-L73

### PR DESCRIPTION
**Pull request by asnare**
This PR updates the TSQL parsing so that the following form of UNION is supported:
```sql
(SELECT a from b)
UNION [ALL]
SELECT x from y
```

Resolves #1127.

### CODE REVIEW

File: core/src/main/scala/com/databricks/labs/remorph/parsers/tsql/TSqlRelationBuilder.scala [hunk 1]
```diff
@@ -66,13 +65,14 @@ class TSqlRelationBuilder(override val vc: TSqlVisitorCoordinator)
...
-        case _ =>
-          // TODO: Implement this style of UNION
-          ir.UnresolvedRelation(
-            ruleText = contextText(ctx),
-            message = s"This style of UNION specification is as yet supported",
-            ruleName = vc.ruleName(ctx),
-            tokenName = Some(tokenName(ctx.getStart)))
+        case qe if qe.queryExpression() != null =>
+          val lhs = qe.queryExpression(0).accept(this)
+          Option(qe.queryExpression(1)).map(_.accept(this)) match {
```
Comment (vil1) [inline core/src/main/scala/com/databricks/labs/remorph/parsers/tsql/TSqlRelationBuilder.scala:70]:
```suggestion
          qe.queryExpression().asScala.lift(1).map(_.accept(this)) match {
```
avoids `IndexOutOfBoundException` if there is only one `queryExpression`

**Review by vil1 — CHANGES_REQUESTED** [review]
could you add a test along the lines of 

```scala
example(
  "(SELECT 1)",
  _.queryExpression(),
  // todo: expected IR
)
```

to see if my concern is relevant?

File: core/src/main/scala/com/databricks/labs/remorph/parsers/tsql/TSqlRelationBuilder.scala (hunk 1 above)
Comment (asnare) [reply core/src/main/scala/com/databricks/labs/remorph/parsers/tsql/TSqlRelationBuilder.scala]: Thanks for this tip: it triggered me to realise I misunderstood something, and now I have a much cleaner version.

File: core/src/test/scala/com/databricks/labs/remorph/parsers/tsql/TSqlRelationBuilderSpec.scala [hunk 2]
```diff
@@ -147,6 +147,40 @@ class TSqlRelationBuilderSpec
...
+            is_all = true,
+            by_name = false,
+            allow_missing_columns = false))
+      }
+      "(SELECT a, b FROM bb)" in {
+        example(
+          "(SELECT a, b FROM bb)",
+          _.queryExpression(),
+          ir.Project(namedTable("bb"), Seq(ir.Column(None, ir.Id("a")), ir.Column(None, ir.Id("b")))))
+      }
```
[the remark this answers is not in the document]
Comment (sundarshankar89) [reply core/src/test/scala/com/databricks/labs/remorph/parsers/tsql/TSqlRelationBuilderSpec.scala:181]: Need to add functional tests, I can show you how it is done.

[the remark this answers is not in the document]
Comment (vil1) [reply core/src/test/scala/com/databricks/labs/remorph/parsers/tsql/TSqlRelationBuilderSpec.scala:181]: @asnare yes, that's the unit test I had in mind

File: core/src/test/scala/com/databricks/labs/remorph/transpilers/TsqlToDatabricksTranspilerTest.scala [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+        "(SELECT a, b from c) UNION SELECT x, y FROM z" -> "(SELECT a, b from c) UNION (SELECT x, y from z);",
+        "(SELECT a, b from c) UNION ALL SELECT x, y FROM z" -> "(SELECT a, b from c) UNION ALL (SELECT x, y from z);",
+        "(SELECT a, b FROM c)" -> "SELECT a, b FROM c;")
+      expectedTranslations.foreach { case (originalTSql, expectedDatabricksSql) =>
+        s"correctly transpile: $originalTSql" in {
+          originalTSql transpilesTo expectedDatabricksSql
+        }
+      }
+    }
+  }
```
Comment (asnare) [inline core/src/test/scala/com/databricks/labs/remorph/transpilers/TsqlToDatabricksTranspilerTest.scala:21]: @sundarshankar89: is this what you had in mind?

Comment (sundarshankar89) [reply core/src/test/scala/com/databricks/labs/remorph/transpilers/TsqlToDatabricksTranspilerTest.scala:21]: under tests/resources/functional/tsql/core_engine/union/

`union_simple.sql`
```
-- tsql sql:
(SELECT a, b from c) UNION SELECT x, y FROM z

-- databricks sql:
(SELECT a, b from c) UNION (SELECT x, y from z);
```
`union_all_simple.sql`
```
-- tsql sql:
(SELECT a, b from c) UNION ALL SELECT x, y FROM z

-- databricks sql:
(SELECT a, b from c) UNION  ALL (SELECT x, y from z);
```

Comment (asnare) [reply core/src/test/scala/com/databricks/labs/remorph/transpilers/TsqlToDatabricksTranspilerTest.scala:21]: @sundarshankar89: Implemented, except in the folder that the python implementation also tests against. (The python implementation already supported this, so the test in that location guards against regressions.)

**Review by vil1 — APPROVED** [review]
LGTM

**Pull request merged**